cpg-analysis/src/test/kotlin/de/fraunhofer/aisec/cpg/passes/UnreachableEOGandDFGTest.kt

```diff
@@ -0,0 +1,215 @@
+ *                   $$ |      $$  ____/ $$ |\_$$ |
+ *                   $$ |  $$\ $$ |      $$ |  $$ |
+ *                   \$$$$$   |$$ |      \$$$$$   |
+ *                    \______/ \__|       \______/
+ *
+ */
+package de.fraunhofer.aisec.cpg.passes
+
+import de.fraunhofer.aisec.cpg.TestUtils
+import de.fraunhofer.aisec.cpg.TranslationManager
+import de.fraunhofer.aisec.cpg.frontends.java.JavaLanguageFrontend
+import de.fraunhofer.aisec.cpg.graph.*
+import de.fraunhofer.aisec.cpg.graph.declarations.FunctionDeclaration
+import de.fraunhofer.aisec.cpg.graph.declarations.RecordDeclaration
+import de.fraunhofer.aisec.cpg.graph.declarations.TranslationUnitDeclaration
+import de.fraunhofer.aisec.cpg.graph.statements.CompoundStatement
+import de.fraunhofer.aisec.cpg.graph.statements.IfStatement
+import java.nio.file.Path
+import kotlin.test.assertEquals
+import kotlin.test.assertNotNull
+import org.junit.jupiter.api.BeforeAll
+import org.junit.jupiter.api.Test
+import org.junit.jupiter.api.TestInstance
+
+@TestInstance(TestInstance.Lifecycle.PER_CLASS)
+class UnreachableEOGandDFGTest {
+
+    private lateinit var tu: TranslationUnitDeclaration
+
+    @BeforeAll
+    fun beforeAll() {
+        val topLevel = Path.of("src", "test", "resources", "passes", "unreachable")
+        TranslationManager.builder().build().analyze()
+        tu =
+            TestUtils.analyzeAndGetFirstTU(
+                listOf(topLevel.resolve("Unreachability.java").toFile()),
+                topLevel,
+                true
+            ) {
+                it.registerLanguage(
+                        JavaLanguageFrontend::class.java,
+                        JavaLanguageFrontend.JAVA_EXTENSIONS
+                    )
+                    .registerPass(UnreachableEOGPassBeforeDFG())
+            }
+    }
+
+    @Test
+    fun testIfTrue() {
+        val method =
+            tu.getDeclarationsByName("TestClass", RecordDeclaration::class.java)
+                .firstOrNull()
+                ?.declarations
+                ?.firstOrNull { d -> d.name == "ifTrue" } as FunctionDeclaration?
+
+        assertNotNull(method)
+
+        val ifStatement = (method.body as CompoundStatement).statements[2] as? IfStatement
+        assertNotNull(ifStatement)
+        val thenY = ifStatement.thenStatement.refs["y"]
+        assertNotNull(thenY)
+
+        val argY = method.calls["println"]?.arguments?.firstOrNull()
+        assertNotNull(argY)
+        // There's only one DFG to the y++ in the then branch.
+        assertEquals(setOf<Node>(thenY), argY.prevDFG)
+    }
+
+    @Test
+    fun testIfFalse() {
+        val method =
+            tu.getDeclarationsByName("TestClass", RecordDeclaration::class.java)
+                .firstOrNull()
+                ?.declarations
+                ?.firstOrNull { d -> d.name == "ifFalse" } as FunctionDeclaration?
+
+        assertNotNull(method)
+
+        val ifStatement = (method.body as CompoundStatement).statements[2] as? IfStatement
```

Comment:
there is this `bodyOrNull` helper function which basically returns the `nth` statement of a certain class. This might help with the weird casts.